### PR TITLE
[MRESOLVER-502] Install/Deploy at session end

### PR DESCRIPTION
**Pull request by cstamas**
With session lifecycle, resolver can now perform "install/deployAtEnd" session-wise: it merely withholds install/deploy requests and executes them at session end.

---

https://issues.apache.org/jira/browse/MRESOLVER-502

### CODE REVIEW

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java [hunk 1]
```diff
@@ -83,6 +86,30 @@
...
+        }
+
+        @Override
+        public void run() {
+            try (SyncContext syncContext = syncContextFactory.newInstance(session, false)) {
+                for (DeployRequest request : requests) {
+                    deploy(syncContext, session, request);
+                }
+            } catch (DeploymentException e) {
+                logger.warn("Failure during deferred deployment:", e);
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java:107]: Colon is redundant

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultInstaller.java [hunk 2]
```diff
@@ -64,6 +68,30 @@
...
+        }
+
+        @Override
+        public void run() {
+            try (SyncContext syncContext = syncContextFactory.newInstance(session, false)) {
+                for (InstallRequest request : requests) {
+                    install(syncContext, session, request);
+                }
+            } catch (InstallationException e) {
+                logger.warn("Failure during deferred installation:", e);
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultInstaller.java:89]: same here

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java (hunk 1 above)
Comment (cstamas) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultDeployer.java]: fixed

**Review by michael-o — APPROVED** [review]
No objections, let's wait for the others.

**Review by gnodet — COMMENTED** [review]
What's the benefit compared to do that in the plugin ?
I don't really see any, however, I do see a drawback which is the fact that the exception which may happen won't be caught and processed in the usual way by maven.

**Comment by cstamas** [comment]
Benefits are manifold, but let me turn around the thing: I don't see any benefit in "failing the build/processing in usual way" by maven. In fact, there is only one (easily fixable) aspect where I do agree: the error output written to console. But in both case, user will have to intervene, as in case of install (usually disk full, or some permission error) or deploy (wrong URL, auth or simply network outage) in both cases user must intervene somehow (free disk space, fix perms, fix URL and/or auth, etc).

OTOH, the real benefit is that install/deploy flow already goes thru Resolver, and now that resolver became aware when session ends, I really see this as "natural fit" to place this logic. Just look at install/deploy plugins how convoluted is their logic (or Takari logic that uses LifecycleParticipant to achieve same). It just makes things natural and a LOT simpler. 

And the other huge benefit is that resolver can, if needed "rearrange" and better parallelize deploys even (not done). And last but not least, generators, that may handle some sort of "time limited session" can be invoked in rather snappy way (although current PR contains "simplistic" deferred implementation, that can be improved in this regard).

To me, this really just moves things to their "proper place".

And about error handling, IMO Maven core would need anyway some handling of errors that may happen on session or system close, as currently we have none of those in place (but baby steps).

**Comment by gnodet** [comment]
> Benefits are manifold, but let me turn around the thing: I don't see any benefit in "failing the build/processing in usual way" by maven. In fact, there is only one (easily fixable) aspect where I do agree: the error output written to console. But in both case, user will have to intervene, as in case of install (usually disk full, or some permission error) or deploy (wrong URL, auth or simply network outage) in both cases user must intervene somehow (free disk space, fix perms, fix URL and/or auth, etc).
> 
> OTOH, the real benefit is that install/deploy flow already goes thru Resolver, and now that resolver became aware when session ends, I really see this as "natural fit" to place this logic. Just look at install/deploy plugins how convoluted is their logic (or Takari logic that uses LifecycleParticipant to achieve same). It just makes things natural and a LOT simpler.
> 
> And the other huge benefit is that resolver can, if needed "rearrange" and better parallelize deploys even (not done). And last but not least, generators, that may handle some sort of "time limited session" can be invoked in rather snappy way (although current PR contains "simplistic" deferred implementation, that can be improved in this regard).
> 
> To me, this really just moves things to their "proper place".

I'm not really convinced.  I don't think the logic in the resolver is very different from the one in the install/deploy plugin, and parallelisation can be done when submitting the big install/deploy request to the resolver.  Anyway, no big deal...

> And about error handling, IMO Maven core would need anyway some handling of errors that may happen on session or system close, as currently we have none of those in place (but baby steps).

Yes, this is a very important aspect imho. The error needs to properly get back to Maven, so that it can be handled correctly.  The build needs to be flagged as not successful with proper reporting, etc..., especially with `mvnd` and a remote process for example.

**Comment by cstamas** [comment]
> Another question, given this is a global flag on the session, how can the install/deploy plugin set it ?

Unsure I get the question, so let me ask back: why would this be set by those plugins? Or in other words, how do you "install/deploy at end" today? By setting a user properties that IS set globally on session at Maven boot. Am really unsure why would it needed to be set by the plugin at all...

**Comment by cstamas** [comment]
> I'm not really convinced. I don't think the logic in the resolver is very different from the one in the install/deploy plugin

Is a LOT different. Here, it is really plain straight forward, while in plugins it does hoops and loops to:
* figure out "is it end" (w/ respect of multi module builds, parallel builds, etc)?
* figure modules in reactor that may not use install plugin and consider them as well
* keep state and do something wildly different if "not end"
* etc.

Here, Installer/Deployer simply gets the "done done" list of artifacts that are to be installed. And done.

And plugins just "install/deploy", no logic, they are no brainer, and IMHO corresponding logic (here in this PR) is also much much simpler. 

Btw, it is a sign that "deploy at end" was not solved properly for so long...

**Comment by cstamas** [comment]
While all this is nice and dandy, I agree with @gnodet 
* this PR clearly shows signs of "being tailored" for current Maven3 use case
* what consumer project should do instead, is to itself gather and create "at end" (potentially big for better parallelization) requests instead
* install/deploy is too critical to be done "at end", just before closing session

**Pull request closed without merging**